ring/partition_ring_test.go

```diff
@@ -143,6 +149,775 @@ func BenchmarkPartitionRing_ActivePartitionForKey(b *testing.B) {
+			},
+		},
+		"shard size < number of partitions": {
+			partitionTokens: map[int32][]uint32{
+				0: {userToken(userID, "", 0) + 1},
+				1: {userToken(userID, "", 1) + 1},
+				2: {userToken(userID, "", 2) + 1},
+			},
+			timeline: []event{
+				{what: add, partitionID: 0, partitionDesc: generatePartitionWithInfo(PartitionActive, outsideLookback)},
+				{what: add, partitionID: 1, partitionDesc: generatePartitionWithInfo(PartitionActive, outsideLookback)},
+				{what: add, partitionID: 2, partitionDesc: generatePartitionWithInfo(PartitionActive, outsideLookback)},
+				{what: test, shardSize: 1, expected: []int32{0}},
+			},
+		},
+		"new partitions added": {
+			partitionTokens: map[int32][]uint32{
+				0: {userToken(userID, "", 0) + 1},
+				1: {userToken(userID, "", 2) + 1},
+				2: {userToken(userID, "", 1) + 1},
+			},
+			timeline: []event{
+				{what: add, partitionID: 0, partitionDesc: generatePartitionWithInfo(PartitionActive, outsideLookback)},
+				{what: add, partitionID: 1, partitionDesc: generatePartitionWithInfo(PartitionActive, outsideLookback)},
+				{what: test, shardSize: 2, expected: []int32{0, 1}},
+				{what: add, partitionID: 2, partitionDesc: generatePartitionWithInfo(PartitionActive, withinLookback)},
+				{what: test, shardSize: 2, expected: []int32{0, 1, 2}},
+				{what: update, partitionID: 2, partitionDesc: generatePartitionWithInfo(PartitionActive, outsideLookback)},
+				{what: test, shardSize: 2, expected: []int32{0, 2}},
+			},
+		},
+		"shard size = 2, partition is marked as inactive then marked as active again": {
+			partitionTokens: map[int32][]uint32{
+				0: {userToken(userID, "", 1) + 1},
+				1: {userToken(userID, "", 2) + 1},
+				2: {userToken(userID, "", 0) + 1},
+			},
+			timeline: []event{
+				{what: add, partitionID: 0, partitionDesc: generatePartitionWithInfo(PartitionActive, outsideLookback)},
+				{what: add, partitionID: 1, partitionDesc: generatePartitionWithInfo(PartitionActive, outsideLookback)},
+				{what: add, partitionID: 2, partitionDesc: generatePartitionWithInfo(PartitionActive, outsideLookback)},
+				{what: test, shardSize: 2, expected: []int32{2, 0}},
+				{what: update, partitionID: 2, partitionDesc: generatePartitionWithInfo(PartitionInactive, withinLookback)},
+				{what: test, shardSize: 2, expected: []int32{2, 1, 0}},
+				// Partition 2 still inactive, but now falls outside the lookback period, there's no need to look back to partition 1
+				{what: update, partitionID: 2, partitionDesc: generatePartitionWithInfo(PartitionInactive, outsideLookback)},
+				{what: test, shardSize: 2, expected: []int32{1, 0}},
+				// Partition 2 becomes active again
+				{what: update, partitionID: 2, partitionDesc: generatePartitionWithInfo(PartitionActive, withinLookback)},
+				{what: test, shardSize: 2, expected: []int32{2, 1, 0}},
+				// Partition 2 still active, but now falls outside the lookback period, there's no need to look back to partition 1
```

Comment:
```suggestion
				// Partition 2 still active, but now falls outside the lookback period, there's no need to include to partition 1
```

(comment applies in few more places in subsequent tests too)

Reply:
Done in https://github.com/grafana/dskit/pull/476/commits/42f25898ee5aee2bdf4c0a5bd72b2b4eda900131